marshal/reflect_cache.go

```diff
@@ -0,0 +1,68 @@
+package marshal
+
+import (
+	"reflect"
+	"strings"
+)
+
+var typeCache = map[reflect.Type][]cachedType{}
+
+var nameCache = map[reflect.Type]string{}
+
+var tagCache = map[string]jsonTag{}
+
+type cachedType struct {
+	reflect.StructField
+	JsonTag string
+}
+
+func getTypeFields(structType reflect.Type) []cachedType {
+	if res, ok := typeCache[structType]; ok {
+		return res
+	}
+	res := []cachedType{}
+	for i := 0; i < structType.NumField(); i++ {
+		field := structType.Field(i)
+		res = append(res, cachedType{
+			StructField: field,
+			JsonTag:     field.Tag.Get("json"),
```

Comment:
Re: my other comment; we can parse the JSON tag here, and include the field index.